Add USE_NAMEDTENSOR compilation flag.

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20168 [namedtensor] Add test/test_namedtensor.py
* #20163 Add namedtensor build and test to the CI
* **#20162 Add USE_NAMEDTENSOR compilation flag.**

Sets the NAMEDTENSOR_ENABLED macro (in cpp) and
tools.setup_helpers.env.NAMEDTENSOR_ENABLED.

Test Plan:
- Compile with NAMEDTENSOR_ENABLED=1. Verify that
`torch.__config__.show()` has -DNAMEDTENSOR_ENABLED in the CXX flags.
- Compile without the flag. Verify that `-DNAMEDTENSOR_ENABLED` is not
present in `torch.__config__.show()`.

gh-metadata: pytorch pytorch 20162 gh/zou3519/32/head

Differential Revision: [D15278211](https://our.internmc.facebook.com/intern/diff/D15278211)

zou3519 committed:
Add USE_NAMEDTENSOR compilation flag.

Sets the NAMEDTENSOR_ENABLED macro (in cpp) and
tools.setup_helpers.env.NAMEDTENSOR_ENABLED.

Test Plan:
- Compile with NAMEDTENSOR_ENABLED=1. Verify that
`torch.__config__.show()` has -DNAMEDTENSOR_ENABLED in the CXX flags.
- Compile without the flag. Verify that `-DNAMEDTENSOR_ENABLED` is not
present in `torch.__config__.show()`.

zou3519 committed:
Update on "Add USE_NAMEDTENSOR compilation flag."

Add USE_NAMEDTENSOR compilation flag.

Sets the NAMEDTENSOR_ENABLED macro (in cpp) and
tools.setup_helpers.env.NAMEDTENSOR_ENABLED.

Test Plan:
- Compile with NAMEDTENSOR_ENABLED=1. Verify that
`torch.__config__.show()` has -DNAMEDTENSOR_ENABLED in the CXX flags.
- Compile without the flag. Verify that `-DNAMEDTENSOR_ENABLED` is not
present in `torch.__config__.show()`.

gh-metadata: pytorch pytorch 20162 gh/zou3519/32/head

gchanan commented on tools/build_pytorch_libs.py:
```diff
@@ -175,6 +175,7 @@ def run_cmake(version,
         USE_CUDA=USE_CUDA,
         USE_DISTRIBUTED=USE_DISTRIBUTED,
         USE_FBGEMM=not (check_env_flag('NO_FBGEMM') or check_negative_env_flag('USE_FBGEMM')),
+        NAMEDTENSOR_ENABLED=(check_env_flag('USE_NAMEDTENSOR') or check_negative_env_flag('NO_NAMEDTENSOR')),
```
why do you need to check the env flag here when other examples (e.g. USE_MKLDNN) don't need to?

zou3519 replied:
@gchanan the other examples do, somewhere down the line. Here is where the USE_MKLDNN flag is defined, for example: https://github.com/pytorch/pytorch/blob/bc5398451e5de266767120a18f75e08f06d4bf8e/tools/setup_helpers/env.py#L63.

gchanan replied:
ok, I don't know why we don't do things in one place, but if it works, it works!

zou3519 committed:
Update on "Add USE_NAMEDTENSOR compilation flag."

Add USE_NAMEDTENSOR compilation flag.

Sets the NAMEDTENSOR_ENABLED macro (in cpp) and
tools.setup_helpers.env.NAMEDTENSOR_ENABLED.

Test Plan:
- Compile with NAMEDTENSOR_ENABLED=1. Verify that
`torch.__config__.show()` has -DNAMEDTENSOR_ENABLED in the CXX flags.
- Compile without the flag. Verify that `-DNAMEDTENSOR_ENABLED` is not
present in `torch.__config__.show()`.

gh-metadata: pytorch pytorch 20162 gh/zou3519/32/head
